onnxruntime/contrib_ops/cpu/skip_layer_norm.cc

```diff
@@ -45,18 +45,30 @@ Status SkipLayerNorm<T>::Compute(OpKernelContext* p_ctx) const {
 
   const auto& input_dims = input->Shape().GetDims();
   size_t input_dims_size = input_dims.size();
-  if (input_dims_size != 3 && input_dims_size != 2) {
+  const auto& skip_dims = skip->Shape().GetDims();
+  size_t skip_dims_size = skip_dims.size();
+
+  if (skip_dims_size != 3 && skip_dims_size != 2) {
     return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
-                           "input is expected to have 3 or 2 dimensions, got ", input_dims_size);
+                           "skip is expected to have 3 or 2 dimensions, got ", skip_dims_size);
   }
+  
+  int skip_sequence_length = static_cast<int>(skip_dims[skip_dims_size - 2]);
 
-  int hidden_size = static_cast<int>(input_dims[input_dims_size - 1]);
+  if (input->Shape() != skip->Shape() && skip_dims[0] != 1 && skip_dims[0] != skip_sequence_length) {
```

Comment:
This check is not enough. We need add a condition that last two dimensions are same for input and skip like the following:
```
if (input->Shape() != skip->Shape() && input_dim_size != 3 && (skip_dims[0] != 1 || skip_dim_size != 2) && skip_dims[skip_dim_size-1] != input_dims[input_dim_size-1] && skip_dims[skip_dim_size-2] != input_dims[input_dim_size-2]) {
```

In this way, we can cover the following cases:
(1) input and skip: (batch_size, sequence_length, hidden_size)
(2) input and skip: (batch_size * sequence_length, hidden_size)
(3) input: (batch_size, sequence_length, hidden_size), skip: (1, sequence_length, hidden_size)
(4) input: (batch_size, sequence_length, hidden_size), skip: (sequence_length, hidden_size)

Reply:
Could these work?:

if (skip_dims[skip_dims_size - 2] != input_dims[input_dims_size - 2]) {
    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
                           "sequence length has to be the same for skip and input");
  }

  if (skip_dims[skip_dims_size - 1] != input_dims[input_dims_size - 1]) {
    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
                           "last dimension of skip needs to be same as input");
  }
  if (input->Shape() != skip->Shape() && skip_dims[0] != 1 && skip_dims[0] != skip_sequence_length && input_dims_size != 3) {
    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
                           "skip is expected to have same shape as input or, a batch size of 1 or no batch size when input has 3 dimensions");
  }

Reply:
Actually, for that last check, this looks better:

if ((input->Shape() != skip->Shape()) && ((skip_dims[0] != 1 || skip_dims_size != 2) && input_dims_size != 3)) {
    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
                           "skip is expected to have same shape as input or, a batch size of 1 or no batch size when input has 3 dimensions");
  }

Reply:
That looks good to me.

Reply:
Okay, I see what you're saying. So, combine the first two checks into one and keep the last one as is?

Reply:
Right